Feature/cli

vyasr opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should based on *maint*. -->
<!-- New features should based on *master*. -->

## Description

<!-- Describe your changes in detail. -->

Adds a simple command line entry point for starting a Python interpreter with an open gsd file.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves: #72 

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

Tested manually.

## Change log

<!-- Propose a change log entry. -->
```
Added a command line convenience interface for opening a GSD file.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/gsd/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**GSD Contributor Agreement**](https://github.com/glotzerlab/gsd/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/gsd/blob/master/doc/credits.rst).

joaander commented on gsd/version.py:
```diff
@@ -0,0 +1,11 @@
+# This software is licensed under the BSD 3-Clause License.
+
+
+__version__ = "2.1.1"
```
Add this file to the bumpversion configuration.
Also add it to the list in `CMakeLists.txt` so it is included in the CMake builds for testing.

joaander commented on gsd/hoomd.py:
```diff
@@ -978,13 +980,42 @@ def __exit__(self, exc_type, exc_value, traceback):
         self.file.close()
 
 
+def open_handle(name, mode='rb', schema_version=(1, 4)):
```
Do we need this? Or should we document the `file` attribute of HOOMDTrajectory and get the handle like this?
```
traj = gsd.hoomd.open(...)
handle = traj.file()
```

vyasr replied:
We could document the `file` attribute, in which case I would suggest making it a property.

I think accessing the file directory is definitely useful when you want fast access to specific chunks without the overhead of full file reading. I've mentioned before that I use that feature for fast label evaluation in signac-flow status updates. The only way I know what arguments to pass to `fl.open` for application, schema, and schema_version by looking at `gsd.hoomd.open` (and I only later discovered that some of these arguments were optional), so I think exposing the raw open file handle would be useful. However, I have no problem with recommending that it be accessed through a property of the trajectory.

joaander replied:
Yes, a property would be best for `HOOMDTrajectory.file`. Exposing this would remove the need for `hoomd.open_handle` and reduce duplication of APIs: `fl.open`, `hoomd.open_handle`, and `hoomd.open`.

To deal with the performance issue you mention, we could refactor the `HOOMDTrajectory` object to lazily load data only when requested. That would require a significant amount of effort and is beyond the scope of this PR. If anyone is interested, feel free to make an issue.

vyasr replied:
I agree that refactoring is out of scope, I was simply stating that in current master even making the file handle is nontrivial because it requires knowing the appropriate values for the the schema, schema_version, and application arguments. Exposing the `file` as a property would largely address that issue.

b-butler commented on gsd/__main__.py:
```diff
@@ -0,0 +1,124 @@
+            'gsd.fl': sys.modules['gsd.fl'],
+        }
+        additional_attributes = "Number of frames: {}\n".format(len(traj))
+    else:
+        raise ValueError("Unsupported schema.")
```
If we want to support other potential use cases for `

b-butler commented on gsd/__main__.py:
```diff
@@ -0,0 +1,124 @@
+    additional_attributes = "\n"
+
+    if args.schema == 'hoomd':
+        handle = open_handle(args.file, mode=args.mode)
```
We don't need the schema if the mode is `ab` or `rb` (i.e. when we are not creating or overwriting a file). Should we try to use the fl module to open a handle from there, so we don't need to use only the `hoomd` schema? In `wb` or `xb` mode, though we do need the check since we cannot create a file without a schema.

joaander commented on .bumpversion.cfg:
```diff
@@ -5,6 +5,8 @@ tag = False
 
+[bumpversion:file:version.py]
+
 [bumpversion:file:gsd/__init__.py]
```
This line is no longer needed in the bumpversion config.

joaander commented on gsd/__main__.py:
```diff
@@ -0,0 +1,150 @@
+interface for interacting with GSD files. The primary entry point is a single
+command for starting a Python interpreter with a GSD file pre-loaded::
+
+    $ gsd read trajectory.gsd
```
Add documention for the CLI in Sphinx and list the accepted arguments and resulting behavior.

Also add an example using the CLI in hoomd-examples.rst. I think you can use the ipython shell magic in a ..ipython block so the example will execute when the docs build.

vyasr replied:
I did this with an IPython block, but there's a small issue: since this opens an actual interpreter, it freezes the doc build until you send a SIGINT, so I just changed it to print the help. The second-to-last commit uses a console block instead of an ipython block if you want to skip the execution entirely, but I don't see a great way around the interpreter issue.

joaander replied:
OK, let's just go without the ipython block then. It isn't critical for this example output to update with each new doc build and you can provide a better example with a static block.

vyasr commented:
@joaander I've updated the examples and updated the build configuration as needed to fix any errors. This should be ready for merging once you and/or @b-butler approve.

joaander commented on doc/cli.rst:
```diff
@@ -0,0 +1,14 @@
+
+**GSD** provides a command line interface for rapid inspection of files from the command line.
+
+Package contents
```
Remove this heading. The way the docs render, the reader doesn't really know or need to know that it comes from a python package.

joaander commented on doc/index.rst:
```diff
@@ -40,6 +40,7 @@ API to specify the initial condition for a HOOMD-blue simulation or analyze traj
 
     python-api
     c-api
+    cli
```
Place this before `c-api` as it logically goes with the Python API.

joaander commented on gsd/__main__.py:
```diff
@@ -0,0 +1,164 @@
+
+.. option:: -m mode, --mode mode
+
+    The mode in which to open the file. Valid modes are identical to
```
```suggestion
    The mode in which to open the file. Valid modes are identical to those accepted by
```
